tests/integration-tests/framework/credential_providers.py

```diff
@@ -54,37 +54,49 @@ def aws_credential_provider(region, credential_arg):
 @contextmanager
 def sts_credential_provider(region, credential_arn, credential_external_id=None, credential_endpoint=None):
     """Assume a role through STS and set such creds in the environment for the duration of the context."""
-    credentials_to_backup = _get_current_credentials()
-
-    logging.info("Assuming STS credentials for region %s and role %s", region, credential_arn)
-    aws_credentials = _retrieve_sts_credential(region, credential_arn, credential_external_id, credential_endpoint)
-    logging.info("Retrieved credentials %s", obfuscate_credentials(aws_credentials))
-
-    try:
-        logging.info("Unsetting current credentials %s", obfuscate_credentials(credentials_to_backup))
-        _unset_credentials()
-
-        os.environ["AWS_ACCESS_KEY_ID"] = aws_credentials["AccessKeyId"]
-        os.environ["AWS_SECRET_ACCESS_KEY"] = aws_credentials["SecretAccessKey"]
-        os.environ["AWS_SESSION_TOKEN"] = aws_credentials["SessionToken"]
-        os.environ["AWS_CREDENTIAL_EXPIRATION"] = aws_credentials["Expiration"].isoformat()
-        boto3.setup_default_session()
-
-        yield aws_credentials
-    finally:
-        logging.info("Restoring credentials %s", obfuscate_credentials(credentials_to_backup))
-        _restore_credentials(credentials_to_backup)
-
-
-# TODO: we could add caching but we need to refresh the creds if 1h is passed or increase the MaxSessionDuration
-def _retrieve_sts_credential(region, credential_arn, credential_external_id=None, credential_endpoint=None):
     if credential_endpoint:
         match = re.search(r"https://sts\.(.*?)\.", credential_endpoint)
         endpoint_region = match.group(1)
         sts = boto3.client("sts", region_name=endpoint_region, endpoint_url=credential_endpoint)
     else:
         sts = boto3.client("sts", region_name=region)
+    caller_arn = sts.get_caller_identity()["Arn"]
+    credentials_to_backup = _get_current_credentials()
+    caller_arn_split = re.split(":|/", caller_arn)
+    credential_arn_split = re.split(":|/", credential_arn)
+    if caller_arn_split[4] == credential_arn_split[4] and caller_arn_split[6] == credential_arn_split[6]:
+        # If account id and role name are the same, the role do not need to / cannot assume itself
+        # Caller ARN can be a IAM role arn or STS session ARN.
+        # Therefore, we have to check the account id and role name are the same, instead of checking the whole string.
+        # For example: User:
+        #   arn:aws:sts::123456700000:assumed-role/JenkinsStsRole-il-central-1/il-central-1_integration_tests_session
+        #   is not authorized to perform: sts:AssumeRole on resource:
+        #   arn:aws:iam::123456700000:role/JenkinsStsRole-il-central-1
```

Comment:
Do we need to check the Expiry of that role? which identity does it take if the assumed role has expired? we might get an exception if it has expired

Reply:
It is nice to have, but currently cannot be done:
1. There is no boto3 calls for checking the expiry time
2. We are not using this function to renew the expiry time intentionally. So I believe this change does not introduce any regression